### PR TITLE
importccl: remove IMPORT formats from grammar

### PR DESCRIPTION
**Pull request by madelynnblue**
This allows us to add new formats without changing the grammar rules.

Release note: None

### CODE REVIEW

**Comment by dt** [comment]
Hm. Can you expand on the motivation? Is editing the grammar proving to be a sticking point / are we adding that many more formats?

I was thinking we wanted todo almost the opposite -- that we should be moving more of the options *into* the grammar, so that e.g. the csv-specific options are only valid to CSV format, mysql in mysql statements, etc. I like that this would also include them in the relevant help/diagrams/tab-completion/etc as that is parser-generated and lets us type params individually.

**Comment by mjibson** [comment]
Raphael asked for it and I liked the idea? I have no horse in this pond or duck in this race so I don't care one way or the other. @knz can you give some motivation for your request?

**Comment by dt** [comment]
related: https://github.com/cockroachdb/cockroach/issues/26150

File: pkg/ccl/importccl/exportcsv.go [hunk 1]
```diff
@@ -72,8 +72,7 @@ func exportPlanHook(
 	}
 
 	if exportStmt.FileFormat != "CSV" {
-		// not possible with current parser rules.
-		return nil, nil, nil, errors.Errorf("unsupported import format: %q", exportStmt.FileFormat)
+		return nil, nil, nil, errors.Errorf("unsupported export format: %q", exportStmt.FileFormat)
```
Comment (knz) [inline pkg/ccl/importccl/exportcsv.go]: I think this is an unrelated change that deserves a separate commit (either in this pr or another)

**Comment by knz** [comment]
Github ate half of my previous comment, so re-writing here:

>  Is editing the grammar proving to be a sticking point / are we adding that many more formats?

Every new keyword slows down the scanner a tiny little bit for every token, with an overhead that spreads over the entire grammar. Also adding new variants to a non-terminal increases overhead of that non-terminal in the parser, and overall memory usage for the parser Of course this is an argument that also denounces some other places in the grammar where we've been a bit prolofic, but generally I think it's good hygiene to not rely on a LALR parser to do everything and the kitchen sink.

The rule of thumb is: only ask the parser to check things that cannot be checked elsewhere reasonably, for example when you need the differentiation of separate grammar rules to produce different AST nodes. If the syntactic variants produce the same AST node with virtually the same parameters, that's an indication the check can be done during semantic analysis instead.

An example of this is whether aggregate or window functions are allowed or not in specific scalar expressions: they should always be rejected in the LIMIT clause, and accepted in the SELECT clause. We could have separate `expr` non-terminal trees, one that accepts aggregates/window functions and another that doesn't, and use either non-terminal in the different clauses that need expressions. But we don't do that, to keep the parser table smaller, and instead we reject the invalid syntax during semantic analysis.

Even when different AST nodes are mandated there are situations where it's so much cheaper to do it in the action code than using separate grammar rules (because the grammar would otherwise be complicated) that we do it this way, check for example the non-terminal `generic_set` that chooses between `SetVar` and `SetTracing`.

> [...]  moving more of the options into the grammar, so that e.g. the csv-specific options are only valid to CSV format, mysql in mysql statements, etc. 

Again these checks can be done during semantic analysis. To bounce on your explanation yesterday, I think that the semantic analysis can be the place where you check that the list of column definitions is explicitly provided for those formats that need it. 

> I like that this would also include them in the relevant help/diagrams

You can enunciate the variants side-by-side in `diagrams.go` explicitly.

> /tab-completion/etc 

Here I am almost confident that we actually do not want the tab completion contextual help to be different for every IMPORT variant, and instead have a single help text for IMPORT that lists all the variants side-by-side. This is because having them side-by-side is more useful to a user navigating/reading contextual help without knowing exactly what they need in advance: they could be invoking the help using one of the syntax variants, and needing to read the other variants to discover that they needed something else. If we split the help texts, they would not be able to organically discover the variants as easily.

File: pkg/ccl/importccl/exportcsv.go (hunk 1 above)
Comment (mjibson) [reply pkg/ccl/importccl/exportcsv.go]: moved to a second commit in this pr

**Comment by mjibson** [comment]
Thanks for the detailed explanation. I'm convinced. PTAL.

File: pkg/sql/parser/sql.y [hunk 2]
```diff
@@ -1644,19 +1625,19 @@ import_data_format:
 //
 // %SeeAlso: CREATE TABLE
 import_stmt:
- IMPORT import_bundle_format '(' string_or_placeholder ')'
+ IMPORT import_format '(' string_or_placeholder ')'
   {
     $$.val = &tree.Import{Bundle: true, FileFormat: $2, Files: tree.Exprs{$4.expr()}}
   }
-| IMPORT TABLE table_name FROM import_bundle_format '(' string_or_placeholder ')'
```
Comment (dt) [inline pkg/sql/parser/sql.y]: nit: there's at least one spot in `import_stmt.go` where we say that they parser won't let a non-bundle format show up with`importStmt.Bundle == true` which is now possible

**Review by dt — APPROVED** [review]

File: pkg/sql/parser/sql.y (hunk 2 above)
Comment (mjibson) [reply pkg/sql/parser/sql.y]: Found and removed.
